### ISSUE TITLE
does mean_count_loss(utilize loss) really work?

### ISSUE THREAD

**Issue by Miopha** [wds1996/GNPR-SID#2]
The`mean_count_loss` in vq.py corresponds to the utilization loss in the paper.
`mean_count_loss` is calculated by the following code
1.
`
indices = torch.argmax(Q, dim=-1)
`
2. 
`
sample_counts = torch.bincount(indices.view(-1), minlength=self.n_e) 
`
`
mean_count = indices.size(0) / self.n_e
`
`
mean_count_loss = torch.mean((sample_counts - mean_count) ** 2) / (mean_count ** 2 + 1e-5)
`
`torch.argmax` is non-differentiable in PyTorch, which blocks the gradient flow of `mean_count_loss`, ultimately resulting in no gradients being produced during the backward pass of`mean_count_loss`.
In other words, during debugging, all the `requires_grad` fields of the above variables are **False** (except `Q`), so I’m really wondering if the loss is actually working?
I'm not sure if I might have missed anything, so I would appreciate any advice.

**Comment by wds1996** [comment]
We refer to other projects for the work of rqvae, please refer to them for details 
`https://github.com/Jyonn/RQ-VAE` 
or 
`https://github.com/EdoardoBotta/RQ-VAE-Recommender`

**Comment by Miopha** [comment]
but this loss was proposed in GNPR, which ensures the uniqueness of SID and avoids semantic collisions (referring from the paper), not in the original RQ-VAE.

**Comment by wds1996** [comment]
I have checked my parameter settings, we did not use the `use_stnear = 1` during training, so this part of the code `indices = torch.argmax(Q, dim=-1)` was not executed.

I haven't seen the code for GNPR, so I'm not sure if the designs for these two losses are consistent.

**Comment by wds1996** [comment]
I have checked the code. The `indices` you mentioned here are only used later to generate indices for coodbooks and are not involved in the subsequent gradient computation.

**Comment by Miopha** [comment]
I really appreciate your timely reply. Could you please tell me which part of the code corresponds to the utilization loss mentioned in the paper?
